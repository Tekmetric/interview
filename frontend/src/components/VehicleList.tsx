import { useEffect, useState } from 'react';
import { useNavigate } from "react-router-dom";
import AuthService from '../services/AuthService';
import VehicleService from '../services/VehicleService';
import VehicleType from '../types/VehicleType';
import { stateStr } from '../types/VehicleType';
import Pagination from '../types/Pagination';
import currencyFormatter from '../utils/CurrencyFormatter';
import * as Icon from 'react-bootstrap-icons';
import DeleteVehicleWarning from './DeleteVehicleWarning';
import './VehicleList.css';

export default function VehicleList() {

    let navigate = useNavigate();
    const [vehicles, setVehicles] = useState<VehicleType[]>([]);
    const [deletingVehicle, setDeletingVehicle] = useState<number | null>(null);
    const [pagination, setPagination] = useState<Pagination>(new Pagination(0, 10));

    function queryData(pageNumber: number, pageSize: number) {
        VehicleService.getVehicles(pageNumber, pageSize).then(response => {
            if (response.data.content) {
                setVehicles(response.data.content);
            }

            setPagination({
                empty: response.data.empty,
                first: response.data.first,
                last: response.data.last,
                numberOfElements: response.data.numberOfElements,
                totalElements: response.data.totalElements,
                totalPages: response.data.totalPages,
                pageSize: response.data.pageable.pageSize,
                pageNumber: response.data.pageable.pageNumber
            });
        },
        error => {
            const status = error.response?.status;
            const errorMessage = error.response?.data?.message || error.message || error.toString();
            if (status === 403 || status === 401) {
                navigate("/login", { replace: true });
            }
            console.log(errorMessage);
        });
    }

    useEffect(() => {
        const user = AuthService.getCurrentUser();
        if (user) {
            queryData(0, 10);

        } else {
            navigate("/login", { replace: true });
        }
    }, []);

    function handleNewVehicle() {
        navigate("/vehicle/create", { replace: true });
    }

    function handleEditVehicle(id: number) {
        navigate("/vehicle/edit/" + id , { replace: true });
    }

    function handleDeleteVehicle(id: number, cancelled: false) {
        if (cancelled) {
            setDeletingVehicle(null);
        } else {
            VehicleService.deleteVehicle(id).then(() => {
                setDeletingVehicle(null);
                handlePageClick(pagination.pageNumber);
            });
        }
    }

    function handlePageClick(pageIndex: number) {
        queryData(pageIndex, pagination.pageSize);
    }

    const vehicleRows = vehicles.map(vehicleDto => {
        return <tr key={vehicleDto.id}>
            <td>{stateStr(vehicleDto.state)}</td>
            <td>{vehicleDto.licensePlate}</td>
            <td>{vehicleDto.brand} {vehicleDto.model}</td>
            <td>{vehicleDto.registrationYear}</td>
            <td>{currencyFormatter(vehicleDto.cost)}</td>
            <td>{(new Date(vehicleDto.creationDate)).toLocaleDateString()}</td>
            <td>
                <a href="#" onClick={() => handleEditVehicle(vehicleDto.id)} className="edit" data-toggle="modal">
                    <Icon.PencilFill color="orange" data-toggle="tooltip" title="Edit" />
                </a>
                <a href="#" onClick={() => setDeletingVehicle(vehicleDto.id)} className="delete" data-toggle="modal">
                    <Icon.TrashFill color="red" data-toggle="tooltip" title="Delete" />
                </a>
            </td>
        </tr>;
    });

    const paginationLinks = Array.from(Array(pagination.totalPages).keys()).map((i) => {
        return <li key={i} className={pagination?.pageNumber === i ? 'page-item active' : 'page-item'}>
            <a href="#" onClick={() => handlePageClick(i)} className="page-link">{i + 1}</a>
        </li>
    });

    return (
        <div className="container-xl card">
            <div className="table-responsive">
                <div className="table-wrapper">
                    <div className="table-title">
                        <div className="row">
                            <div className="col-sm-6">
                                <h2>Manage Vehicles</h2>
                            </div>
                            <div className="col-sm-6">
                                <a href="#" className="btn btn-success" onClick={handleNewVehicle}>
                                    <Icon.PlusCircleFill color="white" /> <span>New Vehicle</span>
                                </a>
                            </div>
                        </div>
                    </div>
                    <table className="table table-striped table-hover">
                        <thead>
                            <tr>
                                <th>State</th>
                                <th>License Plate</th>
                                <th>Model</th>
                                <th>Year</th>
                                <th>Cost</th>
                                <th>Creation Date</th>
                                <th>Actions</th>
                            </tr>
                        </thead>
                        <tbody>
                            {vehicleRows}
                        </tbody>
                    </table>

                    {pagination && (
                        <div className="clearfix">
                            <div className="hint-text">
                                Showing&nbsp;
                                <b>{pagination.pageNumber * pagination.pageSize}</b> to&nbsp; 
                                <b>{pagination.pageNumber * pagination.pageSize + pagination.numberOfElements}</b> out of&nbsp; 
                                <b>{pagination.totalElements}</b> entries
                            </div>
                            <ul className="pagination">
                                <li className={pagination.first ? 'page-item disabled' : 'page-item'}>
                                    <a href="#" onClick={() => handlePageClick(pagination.pageNumber - 1)} className="page-link">&#60;</a>
                                </li>
                                {paginationLinks}
                                <li className={pagination.last ? 'page-item disabled' : 'page-item'}>
                                    <a href="#" onClick={() => handlePageClick(pagination.pageNumber + 1)} className="page-link">&#62;</a>
                                </li>
                            </ul>
                        </div>
                    )}
                </div>
            </div>
            <DeleteVehicleWarning vehicleId={deletingVehicle} onDelete={handleDeleteVehicle} />
        </div>
    );
}