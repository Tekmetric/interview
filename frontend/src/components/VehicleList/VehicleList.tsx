import styles from './vehicleList.module.css';
import { Vehicle } from '../../features/vehicles/types';

interface VehicleListProps {
  vehicles: Vehicle[];
}

const vehicleList = (vehicles: Vehicle[]) => {
  return vehicles.map((vehicle) => {
    return (
      <tr key={vehicle.id}>
        <td>{vehicle.vin}</td>
        <td>{vehicle.make}</td>
        <td>{vehicle.model}</td>
        <td>{vehicle.modelYear}</td>
      </tr>
    );
  });
};

const VehicleList = (props: VehicleListProps) => {
  const { vehicles } = props;

  if (vehicles.length === 0) return <p>No Vehicles Inventoried</p>;

  return (
    <table className={styles.table}>
      <thead>
        <tr>
          <th>vehicle name</th>
          <th>vehicle make</th>
          <th>vehicle model</th>
          <th>vehicle year</th>
        </tr>
      </thead>
      <tbody>{vehicleList(vehicles)}</tbody>
    </table>
  );
};

export default VehicleList;
