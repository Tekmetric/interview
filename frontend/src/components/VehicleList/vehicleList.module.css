.table {
}

.table tr {
  background: #ccc;
}

.table td,
th {
  padding: var(--space-sm) var(--space-xs);
}
.table tr:nth-child(even) {
  background: #f8f8f8;
}

.table td,
.table th {
  color: var(--black-color);
}
