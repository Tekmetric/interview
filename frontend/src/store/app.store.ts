import { configureStore } from "@reduxjs/toolkit";
import sessionReducer from './session.store';
import sightingsReducer from './sightings.store';

export const store = configureStore({
  reducer: {
    session: sessionReducer,
    sightings: sightingsReducer,
  }
});

export type AppStore = typeof store;
export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
