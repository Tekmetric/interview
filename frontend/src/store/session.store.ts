import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { defaultTheme } from "../constants/theme.constants";
import { Themes } from "../types/Theme";

export interface SessionState {
  username: string | undefined,
  isAuthenticated: boolean,
  theme: Themes,
};

const initialState: SessionState = {
  username: undefined,
  isAuthenticated: false,
  theme: defaultTheme
};

export const sessionSlice = createSlice({
  name: 'session',
  initialState,
  reducers: {
    setTheme: (state, action: PayloadAction<Themes>) => {
      state.theme = action.payload;
    },
    completeLogin: (state, action: PayloadAction<string>) => {
      state.username = action.payload;
      state.isAuthenticated = !!action.payload;
    },
    logout: (state) => {
      state.username = undefined;
      state.isAuthenticated = false;
    }
  }
})

export const { setTheme, completeLogin, logout } = sessionSlice.actions;

export default sessionSlice.reducer
