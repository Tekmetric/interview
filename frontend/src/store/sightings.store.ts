import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { Sighting } from "../types/Sighting";

export const DEFAULT_ITEMS_PER_PAGE = 10;

export interface SightingState {
  sightings: Sighting[],
  currentPage: number;
  itemsPerPage: number;
  totalPages: number;
};

const initialState: SightingState = {
  sightings: [],
  currentPage: 0,
  itemsPerPage: DEFAULT_ITEMS_PER_PAGE,
  totalPages: 0,
};

export const sightingSlice = createSlice({
  name: 'sighting',
  initialState,
  reducers: {
    setSightings: (state, action: PayloadAction<Sighting[]>) => {
      state.sightings = action.payload;
    },
    setTotalPages: (state, action: PayloadAction<number>) => {
      state.totalPages = action.payload;
    },
    previousPage: (state) => {
      state.currentPage = Math.max(0, state.currentPage - 1);
    },
    nextPage: (state) => {
      state.currentPage = Math.min(state.totalPages -1, state.currentPage + 1);
    }
  }
})

export const { setSightings, setTotalPages, nextPage, previousPage } = sightingSlice.actions;

export default sightingSlice.reducer
