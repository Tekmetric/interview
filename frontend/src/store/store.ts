import { configureStore } from '@reduxjs/toolkit';
import { Reducer, UnknownAction } from 'redux';
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import { ApplicationState } from '../features/application/types';
import { VehiclesState } from '../features/vehicles/types';
import vehicleReducer from '../features/vehicles/reducer';
import applicationReducer from '../features/application/reducer';

// Casting reducers to fix RTK type mismatch with classic reducer setup
const store = configureStore({
  reducer: {
    application: applicationReducer as Reducer<ApplicationState, UnknownAction>,
    vehicles: vehicleReducer as Reducer<VehiclesState, UnknownAction>,
  },
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;

export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
