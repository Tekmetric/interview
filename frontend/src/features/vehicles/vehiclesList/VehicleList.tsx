import { FC, useEffect } from 'react';
import { useSearchParams } from 'react-router-dom';
import ReactPaginate from 'react-paginate';
import { FaChevronLeft, FaChevronRight } from 'react-icons/fa';

import { useAppSelector, useAppDispatch } from '../../../store/store';
import { deleteVehicleById, fetchVehicles, setSelectedVehicle } from '../actions';
import { Vehicle } from '../../../features/vehicles/types';

import Skeleton from '../../../components/Skeleton';
import ToolBar from '../../../components/Toolbar';
import VehicleTable from '../../../components/VehicleTable/VehicleTable';

import styles from './vehicleList.module.css';

interface VehicleListProps {
  vehicleId?: number;
}

const VehicleList: FC<VehicleListProps> = ({ vehicleId }) => {
  const dispatch = useAppDispatch();
  const [searchParams, setSearchParams] = useSearchParams();
  const currentPageFromUrl = parseInt(searchParams.get('page') || '1', 10);

  const { vehicles, loading, error } = useAppSelector((state) => state.vehicles);
  const { data, meta } = vehicles;

  useEffect(() => {
    dispatch(fetchVehicles(currentPageFromUrl));
  }, [dispatch, currentPageFromUrl]);

  const handleSetPagination = (newPage: any) => {
    const selectedPage = newPage.selected + 1;
    setSearchParams({ page: selectedPage.toString() });

    dispatch(fetchVehicles(selectedPage));
  };

  const handleDeleteVehicle = (id: number) => {
    dispatch(deleteVehicleById(id));
  };

  const handleSetSelectedVehicle = (vehicle: Vehicle) => {
    dispatch(setSelectedVehicle(vehicle));
  };

  const tableContent = loading ? (
    <Skeleton count={1} />
  ) : (
    <VehicleTable
      vehicles={data}
      handleDeleteVehicle={handleDeleteVehicle}
      handleSetSelectedVehicle={handleSetSelectedVehicle}
    />
  );

  const paginationContent = loading ? (
    <Skeleton count={1} />
  ) : (
    <ReactPaginate
      previousLabel={<FaChevronLeft />}
      nextLabel={<FaChevronRight />}
      breakLabel={'...'}
      pageCount={meta.totalPages}
      pageRangeDisplayed={5}
      marginPagesDisplayed={2}
      forcePage={meta.totalPages > 0 ? meta.currentPage - 1 : undefined}
      onPageChange={handleSetPagination}
      containerClassName={'pagination'}
      pageClassName={'page-item'}
      pageLinkClassName={'page-link'}
      activeClassName={'active'}
      disabledClassName={'disabled'}
      breakClassName={'break'}
    />
  );

  return (
    <>
      <div className={styles.container__toolbar}>
        <ToolBar />
      </div>
      <div className={styles.container__table}>{tableContent}</div>
      <div className={styles.container__pagination}>{paginationContent}</div>
    </>
  );
};

export default VehicleList;
