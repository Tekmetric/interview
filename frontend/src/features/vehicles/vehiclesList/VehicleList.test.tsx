import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import store from '../../../store/store';
import VehicleList from './VehicleList';

describe('VehicleList', () => {
  it('renders correctly', () => {
    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <VehicleList />
        </MemoryRouter>
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });
});
