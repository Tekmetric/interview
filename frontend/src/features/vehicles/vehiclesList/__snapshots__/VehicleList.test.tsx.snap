// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VehicleForm > renders correctly 1`] = `
<div>
  <div
    class="_container__toolbar_9bced2"
  >
    <section
      class="_header_e52d52"
    >
      <div
        class="_searchBar_e52d52"
      >
        <a
          class="link-button"
          data-discover="true"
          href="/vehicle/create"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
            />
          </svg>
           
          <span>
            add new vehicle
          </span>
        </a>
        <form
          class="_search_e52d52"
        >
          <input
            data-testid="search"
            placeholder="search"
            type="text"
            value=""
          />
          <button
            aria-label="Search"
            type="submit"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
              />
            </svg>
          </button>
        </form>
      </div>
      <div
        class="_results_e52d52"
        data-id="true"
      >
        <p />
      </div>
    </section>
  </div>
  <div
    class="_container__table_9bced2"
  >
    <p
      class="_empty_03378d"
    >
      No Results Found
    </p>
  </div>
  <div
    class="_container__pagination_9bced2"
  />
</div>
`;

exports[`VehicleList > renders correctly 1`] = `
<div>
  <div
    class="_container__toolbar_9bced2"
  >
    <section
      class="_header_e52d52"
    >
      <div
        class="_searchBar_e52d52"
      >
        <a
          class="link-button"
          data-discover="true"
          href="/vehicle/create"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
            />
          </svg>
           
          <span>
            add new vehicle
          </span>
        </a>
        <form
          class="_search_e52d52"
        >
          <input
            data-testid="search"
            placeholder="search"
            type="text"
            value=""
          />
          <button
            aria-label="Search"
            type="submit"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
              />
            </svg>
          </button>
        </form>
      </div>
      <div
        class="_results_e52d52"
        data-id="true"
      >
        <p />
      </div>
    </section>
  </div>
  <div
    class="_container__table_9bced2"
  >
    <p
      class="_empty_03378d"
    >
      No Results Found
    </p>
  </div>
  <div
    class="_container__pagination_9bced2"
  />
</div>
`;
